tmva/tmva/inc/TMVA/DNN/Adam.h

```diff
@@ -0,0 +1,237 @@
+                                                              const std::vector<Matrix_t> &weightGradients) -> void
+{
+   std::vector<Matrix_t> &currentLayerFirstMomentWeights = this->GetFirstMomentWeightsAt(layerIndex);
+   std::vector<Matrix_t> &currentLayerSecondMomentWeights = this->GetSecondMomentWeightsAt(layerIndex);
+
+   for (size_t k = 0; k < currentLayerFirstMomentWeights.size(); k++) {
+
+      // accumulation matrix used for temporary storing of the current accumulation
+      Matrix_t accumulation(currentLayerFirstMomentWeights[k].GetNrows(), currentLayerFirstMomentWeights[k].GetNcols());
+
+      // Mt = beta1 * Mt-1 + (1-beta1) * currentWeightGradients
+      initialize<Architecture_t>(accumulation, EInitialization::kZero);
+      Architecture_t::ScaleAdd(accumulation, currentLayerFirstMomentWeights[k], this->GetBeta1());
+      Architecture_t::ScaleAdd(accumulation, weightGradients[k], 1 - (this->GetBeta1()));
+      Architecture_t::Copy(currentLayerFirstMomentWeights[k], accumulation);
+
+      // Vt = beta2 * Vt-1 + (1-beta2) * currentSquaredWeightGradients
+      initialize<Architecture_t>(accumulation, EInitialization::kZero);
+      Matrix_t currentSquaredWeightGradients(weightGradients[k].GetNrows(), weightGradients[k].GetNcols());
+      Architecture_t::Copy(currentSquaredWeightGradients, weightGradients[k]);
+      Architecture_t::SquareElementWise(currentSquaredWeightGradients);
+      Architecture_t::ScaleAdd(accumulation, currentLayerSecondMomentWeights[k], this->GetBeta2());
+      Architecture_t::ScaleAdd(accumulation, currentSquaredWeightGradients, 1 - (this->GetBeta2()));
+      Architecture_t::Copy(currentLayerSecondMomentWeights[k], accumulation);
+   }
+
+   // alpha = learningRate * sqrt(1 - beta2^t) / (1-beta1^t)
+   Scalar_t alpha = (this->GetLearningRate()) * (sqrt(1 - pow(this->GetBeta2(), this->GetGlobalStep()))) /
+                    (1 - pow(this->GetBeta1(), this->GetGlobalStep()));
+
+   // updating the weights.
+   // theta = theta - alpha * Mt / (sqrt(Vt) + epsilon)
+   for (size_t i = 0; i < weights.size(); i++) {
+      Matrix_t currentWeightUpdates(weights[i].GetNrows(), weights[i].GetNcols());
+      Architecture_t::Copy(currentWeightUpdates, currentLayerSecondMomentWeights[i]);
+      Architecture_t::SqrtElementWise(currentWeightUpdates);
+      Architecture_t::ConstAdd(currentWeightUpdates, this->GetEpsilon());
+      Architecture_t::ReciprocalElementWise(currentWeightUpdates);
+      Architecture_t::Hadamard(currentWeightUpdates, currentLayerFirstMomentWeights[i]);
+      Architecture_t::ScaleAdd(weights[i], currentWeightUpdates, -alpha);
+   }
+}
+
+//_________________________________________________________________________________________________
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+auto TAdam<Architecture_t, Layer_t, DeepNet_t>::UpdateBiases(size_t layerIndex, std::vector<Matrix_t> &biases,
```

Comment:
update weights and update biases have exactly the same code. can you reduce the code duplication?